Add SdkArchiveDiff task to verify the sdk archive has all the expected files

jtschuster opened the pull request:
Adds a task to validate that the final sdk archive has the same files as the daily installer builds. It will download the latest daily build from the same aka.ms links found in the installer table, so it's not a perfect version match, but it should be accurate enough. It also will only report any differences as messages, it won't fail the build.

Part of the fix for https://github.com/dotnet/source-build/issues/4119

ViktorHofer commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff.csproj:
```diff
@@ -0,0 +1,12 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Library</OutputType>
```
```suggestion
```

ViktorHofer commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff.csproj:
```diff
@@ -0,0 +1,12 @@
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Build.Utilities.Core" PrivateAssets="all" ExcludeAssets="Runtime" Version="$(MicrosoftBuildVersion)" />
+  </ItemGroup>
+</Project>
```
style nit:

```suggestion

</Project>
```

ViktorHofer commented:
> This isn't currently fully running in any CI pipelines since the sdk repo is still excluded in non-sourcebuild VMR builds and sourcebuild only runs on the centos.8-x64 rid. Is there a pipeline that runs the VMR in source-build mode with a portable RID?

No, source-build is non-portable by definition. But we should soon have the sdk repository running in the VMR.

ViktorHofer commented:
Installer now builds live in the unified-build pipeline.

jtschuster commented:
Looks like it found a few differences in the ubuntu release, adding a `.exe` extension to one, and missing a windows(?) dll.

I'll fix the `/` at the end of the file path.

```
  Difference in sdk archive: Removed: sdk/{VERSION}/Containers/containerize/containerize/
  Difference in sdk archive: Added: sdk/{VERSION}/Containers/containerize/containerize.exe/
  Difference in sdk archive: Removed: sdk/{VERSION}/DotnetTools/dotnet-watch/{VERSION}/tools/net9.0/any/runtimes/win/lib/net9.0/System.Diagnostics.EventLog.Messages.dll/
```

mmitche commented:
Interesting! /cc @ViktorHofer

mmitche commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,29 @@
+          AfterTargets="Build"
+          DependsOnTargets="DetermineSourceBuiltSdkVersion" >
+
+      <GetSingleTarballItem SdkTarballItems="@(SdkTarballItem)">
```
Similar for other terminology that may be interchangeable between zip and .tar.gz

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/Archive.cs:
```diff
@@ -0,0 +1,160 @@
+            return _archive.Entries.Select(e => e.FullName).ToArray();
+        }
+
+        public override string[] GetFileLines(string relativePath)
```
`GetFileLines` and `GetFileBytesAsync` are for future use?

jtschuster replied:
Yes, I can remove them for now though to avoid dead code.

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/GetSingleTarballItem.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            try
+            {
+                var (versionString, rid, extension) = Archive.GetInfoFromArchivePath(item.ItemSpec);
```
These are unused.

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/GetSingleTarballItem.cs:
```diff
@@ -0,0 +1,51 @@
+using System.Collections.Generic;
+using Microsoft.Build.Framework;
+
+public class GetSingleTarballItem : Microsoft.Build.Utilities.Task
```
I think maybe the intention is to use this for more, but technically you don't need a task to select a single item from the itemgroup. ItemGroups have intrinsic size functions for checking 1 or 0 cases, and you can use the following to select the first (or last without Reverse). Probably less efficient though.

```
<Target Name="FirstCompile" BeforeTargets="Compile">
  <ItemGroup>
    <Reversed Include="@(Compile->Reverse())" />
  </ItemGroup>
  <PropertyGroup>
    <First>%(Reversed.Identity)</First>
  </PropertyGroup>

  <Message Text="First compile item is $(First)" Importance="high" />
</Target>
```

jtschuster replied:
Yeah, this task tests each item to make sure we select an item from the group that the next task can parse the version and RID info from, not just the first item.

mmitche commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,29 @@
+          AfterTargets="Build"
+          DependsOnTargets="DetermineSourceBuiltSdkVersion" >
+
+      <GetSingleArchiveItem SdkArchiveItems="@(SdkTarballItem)">
```
Instead of a selecting a single item, would it be better to just compare each possible item?

mmitche replied:
And error if the number of possible comparisons is zero

jtschuster replied:
I only needed this because Windows has a .zip, and .msi and .msi.zip in SdkTarballItem, and I needed to choose the .zip. Is there a scenario where there would be multiple valid archives that it would make sense to check multiple?

mmitche replied:
Ahh I see. I would check the .zip in that case, not the wixpack.zip, and explicitly exclude wixpacks from that input set. Other repos may have multiple archives though. Runtime has the main runtime zip, then a few others (targeting pack?)

jtschuster replied:
I changed that task to GetValidArchiveItems that returns all valid items and errors if there are none, then assert afterwards that there's exactly one valid archive.

mmitche commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,29 @@
+      </GetSingleArchiveItem>
+
+      <GetClosestOfficialSdk BuiltSdkPath="$(_BuiltSdkArchivePath)" Condition="'$(_BuiltSdkArchivePath)' != ''">
+        <Output TaskParameter="ClosestOfficialSdkPath" PropertyName="ClosestOfficialSdkPath" />
```
Nit: You use the `_` prefix above for an internal property above, but don't here. I'd use the `_` in both cases.

mmitche commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,29 @@
+      <Message Text="Failed to find closest official SDK archive." Importance="High" Condition="'$(_BuiltSdkArchivePath)' != '' AND '$(ClosestOfficialSdkPath)' == ''" />
+
+      <FindArchiveDiffs BaselineArchive="$(_BuiltSdkArchivePath)" TestArchive="$(ClosestOfficialSdkPath)" Condition="'$(_BuiltSdkArchivePath)' != '' AND '$(ClosestOfficialSdkPath)' != ''">
+        <Output TaskParameter="ContentDifferences" ItemName="ContentDifferences" />
```
Same here

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/Archive.cs:
```diff
@@ -0,0 +1,124 @@
+        var dashDelimitedParts = filename.Split('-');
+        var (rid, versionString) = dashDelimitedParts switch
+        {
+            ["dotnet", "sdk", var first, var second, var third, var fourth] when PathWithVersions.IsVersionString(first) => (third + '-' + fourth, first + '-' + second),
```
Given the complexity of the version numbers that show up in the set of assets produced by .NET, we created a version identifier a long time ago. It's what powers creation of the aka.ms links: https://github.com/dotnet/arcade/blob/main/src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/VersionIdentifier.cs#L51

I'm not totally sure how well it works for paths inside an SDK, but it may be worth looking at if you start to run into corner cases with the checker.

jtschuster replied:
Thanks for the link! I copied that file to this project. I tried referencing that package, but it references a different version of MSBuild that Installer uses and that caused build issues.

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/FindArchiveDiffs.cs:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    public static List<(string, DifferenceKind DifferenceKind)> GetDiffs(
+        string[] originalPathsWithVersions,
```
```suggestion
        string[] baselinePathsWithVersions,
```
?

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/FindArchiveDiffs.cs:
```diff
@@ -0,0 +1,146 @@
+
+    public static List<(string, DifferenceKind DifferenceKind)> GetDiffs(
+        string[] originalPathsWithVersions,
+        string[] modifiedPathsWithVersions,
```
testPathsWithVersions?

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/FindArchiveDiffs.cs:
```diff
@@ -0,0 +1,146 @@
+        Unchanged
+    }
+
+    public static List<(string, DifferenceKind DifferenceKind)> GetDiffs(
```
Would you put a comment blurb up here indicating what the algorithm is that is determining whether there are diffs?

mmitche commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,29 @@
+        <Output TaskParameter="ClosestOfficialSdkPath" PropertyName="ClosestOfficialSdkPath" />
+      </GetClosestOfficialSdk>
+
+      <Message Text="Failed to find closest official SDK archive." Importance="High" Condition="'$(_BuiltSdkArchivePath)' != '' AND '$(ClosestOfficialSdkPath)' == ''" />
```
If the built archive path was empty, wouldn't this also be an error?

jtschuster replied:
Yeah, I reworked some of the errors. We should now error in the tasks if we can't find a valid archive, or if downloading the closest official sdk fails.

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/GetClosestOfficialSdk.cs:
```diff
@@ -0,0 +1,73 @@
+        if (redirectResponse.StatusCode != HttpStatusCode.Moved)
+        {
+            Log.LogMessage(MessageImportance.High, $"Failed to download '{downloadUrl}': invalid aka.ms URL");
+            return true;
```
return false?

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/GetClosestOfficialSdk.cs:
```diff
@@ -0,0 +1,73 @@
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+
+public class GetClosestOfficialSdk : Microsoft.Build.Utilities.Task, ICancelableTask
```
There's a mix of Sdk specific functionality and non-SDK specific functionality here (or at least terminology). For instance, the comparison (find archive diffs) I think works for any archive. On the other hand, Archive has functionality that only works for the SDK (retrieving version numbers).

The below code mostly works for anything, except for `GetLatest`. Maybe it would be better to factor the common code below out into a base task class?

jtschuster replied:
Good idea, I extracted the non-specific functionality to `GetClosestArchive`

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/PathWithVersions.cs:
```diff
@@ -0,0 +1,112 @@
+        ReadOnlySpan<char> directory2 = path2;
+        while (TryGetPathLeaf(directory, out var root, out var directoryPart) && TryGetPathLeaf(directory2, out var root2, out var directoryPart2))
+        {
+            if (!ReplaceVersionString(directoryPart).SequenceEqual(ReplaceVersionString(directoryPart2)))
```
This is one place where that version code in arcade may be helpful. It rips the version elements of the paths out to create  aka.ms links, which is what is happening here.

akoeplinger commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,46 @@
+
+      <Message Text="Difference in sdk archive: %(_ContentDifferences.Kind): %(_ContentDifferences.Identity)"
+               Importance="High"
+               Condition="'%(_ContentDifferences.Identity)' != '' AND '%(_ContentDifferences.Kind)' != 'Unchanged'"/>
```
I'd also add a message when no differences were found

mmitche commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/VersionIdentifier.cs:
```diff
@@ -0,0 +1,257 @@
+using System.Linq;
+using System.Text;
+
+// Copied from https://github.com/dotnet/arcade/blob/903d35c426248ade56719db5bc0947cd7634b297/eng/common/native/init-distro-rid.sh
```
Not ideal, but since this is automation that doesn't need to live forever, I'm fine with it.

mmitche commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,54 @@
+      <Error Text="Multiple valid dotnet-sdk archives found."
+             Condition="'@(_BuiltSdkArchivePath->Count())' != '1'" />
+
+      <GetClosestOfficialSdk BuiltArchivePath="@(_BuiltSdkArchivePath)"
```
I would be surprised if this fails to find a matching SDK...does this happen?

jtschuster replied:
Are you referring to the `Condition="'@(_BuiltSdkArchivePath)' != ''"` or something else? I think we error in the tasks before we'd hit the conditions for `GetClosestOfficialSdk` and `FindArchiveDiffs`, so I'll remove the conditions.

mmitche replied:
Ahh, here: https://github.com/dotnet/installer/pull/18748/files#diff-a4e89432a27e7959cd783637704d06e20a30c549d96948cc59ccea3bb16c7796R33

I would be surprised if the returned `_ClosestOfficialSdkPath` is empty.

jtschuster replied:
Gotcha, yeah that should never happen, we'll error out in the task first. I removed those conditions.

jtschuster commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/GetClosestArchive.cs:
```diff
@@ -0,0 +1,94 @@
+    /// <summary>
+    /// Get the URL of the latest official archive for the given version string and RID.
+    /// </summary>
+    public abstract Task<string?> GetLatestOfficialArchiveUrl();
```
Yeah, that's true, but since ideally this would eventually get the installer commit or parse the build version and find the best build that's not necessarily the latest I'm leaning towards keeping "Closest".

ViktorHofer commented on src/SourceBuild/content/build.proj:
```diff
@@ -20,5 +20,6 @@
   </Target>
 
   <Import Project="$(RepositoryEngineeringDir)build.sourcebuild.targets" Condition="'$(DotNetBuildSourceOnly)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)build.targets" />
```
Please make the targets file name more descriptive, i.e. `sdkarchivediff.targets`.

ViktorHofer commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,52 @@
+<Project>
+  <UsingTask AssemblyFile="$(SdkArchiveDiffTasksAssembly)" TaskName="GetValidArchiveItems" />
```
readability nit:

```suggestion

  <UsingTask AssemblyFile="$(SdkArchiveDiffTasksAssembly)" TaskName="GetValidArchiveItems" />
```

ViktorHofer commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,52 @@
+  <Target Name="ReportSdkArchiveDiffs"
+          AfterTargets="Build"
+          DependsOnTargets="DetermineSourceBuiltSdkVersion"
+          Condition="'$(ShortStack)' != 'true' AND '$(PortableBuild)' == 'true'" >
```
nit: We use lowercase msbuild operators in this repo:

 ```suggestion
          Condition="'$(ShortStack)' != 'true' and '$(PortableBuild)' == 'true'" >
```

ViktorHofer replied:
Move this condition out into the targets file import so that builds that don't need this, don't need to evaluate this file.

ViktorHofer commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,52 @@
+
+      <Delete Files="$(_ClosestOfficialSdkPath)"
+              Condition="'$(_ClosestOfficialSdkPath)' == ''"/>
+
```
```suggestion
```

ViktorHofer commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,52 @@
+               Importance="High"
+               Condition="'@(_changedFiles->Count())' == '0'" />
+
+      <Delete Files="$(_ClosestOfficialSdkPath)"
+              Condition="'$(_ClosestOfficialSdkPath)' == ''"/>
```
This doesn't make sense to me. If the property is empty, delete an empty path file?

jtschuster replied:
Oh that definitely should have been a `!=`, thank you for pointing that out. Though actually we should error earlier if _ClosestOfficialSdkPath isn't set, so we can remove that condition.

ViktorHofer commented on src/SourceBuild/content/eng/build.targets:
```diff
@@ -0,0 +1,52 @@
+                PropertyName="_ClosestOfficialSdkPath" />
+      </GetClosestOfficialSdk>
+
+      <FindArchiveDiffs BaselineArchive="@(_BuiltSdkArchivePath)"
```
I think it would make more sense to create a text file with the diff (i.e. a csv) that we then analyze with a reporting tool to build a dashboard, than adding messages to stdout.

jtschuster replied:
Do you think a csv or a `git diff` style `.diff` file would make more sense?

ViktorHofer replied:
Csv is fine as it would be great to then build a grafana or similar dashboard that takes the Csv as an input.

ViktorHofer commented on src/SourceBuild/content/eng/tools/init-build.proj:
```diff
@@ -116,6 +117,14 @@
     </Touch>
   </Target>
 
+  <Target Name="BuildSdkArchiveDiff" >
```
This target should probably have the same condition as the targets import in build.proj so that the build task only gets built when it is used.

ViktorHofer commented on src/SourceBuild/content/eng/sdkArchiveDiffs.targets:
```diff
@@ -37,16 +36,20 @@
-      <Delete Files="$(_ClosestOfficialSdkPath)"
-              Condition="'$(_ClosestOfficialSdkPath)' == ''"/>
-
+      <Delete Files="$(_ClosestOfficialSdkPath)" />
```
What's the value of `_ClosestOfficialSdkPath` and why do we need to delete the file?

jtschuster replied:
It's the downloaded sdk that we use as a baseline in the comparison. Maybe it should have a clearer name?

ViktorHofer replied:
Where does that get downloaded to? Which component decides where to download that file to?

jtschuster replied:
The `GetClosestSdk` task downloads it to a temp folder that it generates.

ViktorHofer replied:
Do we usually delete files that we put into a temporary directory? Just wondering.

jtschuster replied:
I'm not sure if we usually do or not

ViktorHofer replied:
OK disregard. I think the deletion is fine.

ViktorHofer commented on src/SourceBuild/content/eng/sdkArchiveDiffs.targets:
```diff
@@ -0,0 +1,55 @@
+      </ItemGroup>
+
+      <PropertyGroup>
+        <SdkArchiveDiffsReport>$(ArtifactsLogDir)\SdkArchiveDiffs\SdkArchiveDiffs.csv</SdkArchiveDiffsReport>
```
I don't think we need a sub-folder:

```suggestion
        <SdkArchiveDiffsReport>$(ArtifactsLogDir)SdkArchiveDiffs.csv</SdkArchiveDiffsReport>
```

jtschuster commented on src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.SourceBuild.Tasks.SdkArchiveDiff/GetClosestArchive.cs:
```diff
@@ -0,0 +1,94 @@
+
+        var packageUriPath = packageResponse.RequestMessage!.RequestUri!.LocalPath;
+
+        ClosestOfficialArchivePath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName() + $".{ArchiveName}-{BuiltVersion}-{BuiltRid}.closest.{ArchiveExtension}");
```
Yeah, we could use the filename from the http request instead. Do you think that would make more sense?

ViktorHofer replied:
It probably would be better, yes, as the manual file name construction might not be completely accurate.

ViktorHofer commented on src/SourceBuild/content/eng/sdkArchiveDiffs.targets:
```diff
@@ -0,0 +1,55 @@
+        <SdkArchiveDiffsReport>$(ArtifactsLogDir)SdkArchiveDiffs.csv</SdkArchiveDiffsReport>
+      </PropertyGroup>
+
+      <WriteLinesToFile File="$(SdkArchiveDiffsReport)" Lines="%(_ContentDifference.Kind), %(_ContentDifferences.Identity)" />
```
Add Overwrite="true" and WriteOnlyWhenDifferent="true"

akoeplinger commented on src/SourceBuild/content/eng/sdkArchiveDiff.targets:
```diff
@@ -0,0 +1,64 @@
+        </_sdkFilesDiff>
+        <_sdkFilesDiff Include="@(_ContentDifferences)" Condition="'%(_contentDifferences.Kind)' == 'Unchanged'" >
+          <DiffIndicator> </DiffIndicator>
+        </_sdkFilesDiff>
```
I'm not sure having the Unchanged files is really valuable in the diff, it just makes the file larger than necessary and ideally if there's no diff the file would be empty/non-existing.

akoeplinger commented on src/SourceBuild/content/eng/sdkArchiveDiff.targets:
```diff
@@ -0,0 +1,64 @@
+        </_sdkFilesDiff>
+        <_sdkFilesDiff Include="@(_ContentDifferences)" Condition="'%(_contentDifferences.Kind)' == 'Removed'" >
+          <DiffIndicator>-</DiffIndicator>
+        </_sdkFilesDiff>
```
Checking the Ubuntu run I see:

```
+ templates/{VERSION}/microsoft.dotnet.common.itemtemplates.{VERSION}.nupkg
+ sdk/{VERSION}/Containers/containerize/containerize.exe
- templates/{VERSION}/microsoft.dotnet.common.itemtemplates.{VERSION}.nupkg
- sdk/{VERSION}/Containers/containerize/containerize
- sdk/{VERSION}/DotnetTools/dotnet-watch/{VERSION}/tools/net9.0/any/runtimes/win/lib/net9.0/System.Diagnostics.EventLog.Messages.dll
```

The "templates" one doesn't make sense to me since it seems to be the same?

jtschuster replied:
Yeah, that one is weird. I wonder if it's related to having two Versions in the path? I'll look into it.

akoeplinger commented on src/SourceBuild/content/eng/sdkArchiveDiff.targets:
```diff
@@ -0,0 +1,64 @@
+      </ItemGroup>
+
+      <PropertyGroup>
+        <SdkArchiveDiffsReport>$(ArtifactsLogDir)SdkArchiveContent.diff</SdkArchiveDiffsReport>
```
instead of .diff what if we generate xunit/mstest xml reports instead? those could be picked up by the AzDO test tab so we don't have to build custom UI.
